template.yaml

```diff
@@ -1011,7 +1012,167 @@ Resources:
                   - !Ref AWS::StackId
       RetentionInDays: 7
   
+  GetProductCode:
+    Type: Custom::Lambda
+    Properties:
+      ServiceToken: !GetAtt GetProductCodeCustomResource.Arn
+      ProductId: !Ref ProductId
+      
+  GetProductCodeCustomResource:
+    Type: AWS::Lambda::Function
+    Properties:
+      Role: !GetAtt CAPILambdasExecutionRole.Arn
+      Runtime: nodejs18.x
+      Handler: index.handler
+      Code: 
+        ZipFile: |
+          const { MarketplaceCatalogClient, DescribeEntityCommand } = require("@aws-sdk/client-marketplace-catalog");
+          const response = require('cfn-response');
+          exports.handler = async (event, context) => {
+          context.callbackWaitsForEmptyEventLoop = true;
+            console.log("REQUEST RECEIVED:\n" + JSON.stringify(event));
+            const client = new MarketplaceCatalogClient({ region: 'us-east-1' });
+            const productId = event.ResourceProperties.ProductId; // Assuming the product ID is passed as an event parameter
+
+            try {
+                if (event.RequestType === 'Create' || event.RequestType === 'Update') {
+                    // Fetch the product details from AWS Marketplace
+                    const command = new DescribeEntityCommand({
+                        Catalog: 'AWSMarketplace',
+                        EntityId: productId,
+                        EntityType: 'Product'
+                    });
+                    const resp = await client.send(command);
+
+                    // Extract the product code from the response
+                    const productCode = resp.DetailsDocument.Description.ProductCode;
+
+                    const responseData = {
+                        ProductCode: productCode
+                    };
+
+                    await response.send(event, context, 'SUCCESS', responseData);
+                } else if (event.RequestType === 'Delete') {
+                    // No action needed for delete
+                    await response.send(event, context, 'SUCCESS', {});
+                } else {
+                    await response.send(event, context, 'FAILED', { error: 'Invalid request type' });
+                }
+            } catch (error) {
+                console.error('Error:', error);
+                await response.send(event, context, 'FAILED', { error: 'Failed to fetch product code' });
+            }
+          };
+
+  CAPILambdasExecutionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Version: '2012-10-17'
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service:
+                - lambda.amazonaws.com
+            Action:
+              - 'sts:AssumeRole'
+      ManagedPolicyArns:
+        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
+        - arn:aws:iam::aws:policy/AWSMarketplaceSellerFullAccess
```

Comment:
This managed policy is too wide: "AWSMarketplaceSellerFullAccess"

We can use resource level conditions to scope it down.
https://docs.aws.amazon.com/marketplace-catalog/latest/api-reference/api-access-control.html#allowing-actions-on-specific-resources

Here is how (untested):
```
  CAPILambdasExecutionRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Version: 2012-10-17
        Statement:
          - Effect: Allow
            Principal:
              Service:
                - lambda.amazonaws.com
            Action:
              - "sts:AssumeRole"
      Policies:
        - PolicyName: manage-products
          PolicyDocument:
            Version: 2012-10-17
            Statement:
              - Effect: Allow
                Action:
                  - "aws-marketplace:StartChangeSet"
                  - "aws-marketplace:DescribeEntity"
                Resource:
                  - !Sub "arn:${AWS::Partition}:aws-marketplace:us-east-1:${AWS::AccountId}:AWSMarketplace/SaaSProduct/${ProductId}"
                  - !Sub "arn:${AWS::Partition}:aws-marketplace:us-east-1:${AWS::AccountId}:AWSMarketplace/ChangeSet/*"
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
```